File: CA4/3.cpp
//https://chatgpt.com/share/676d9f44-7604-800e-a442-848b2f491c24
//Sayyed Mohammad Hossein Mazhari
//CA4 Q3
#include <iostream>
#include <vector>
#include <string>

using namespace std;

typedef long long ll;
struct Edge { int node1; int node2; int index; };
struct Adj { int node; int index; };

const int MAX_WEIGHT = 1e6 + 7;
const int MAX_NODES = 1e5 + 7;

int findParent(int node, vector<int>& parent) {
    if (node == parent[node]) {
        return node;
    }
    return parent[node] = findParent(parent[node], parent);
}

void depthFirstSearch(int node, int callerIndex, int time, vector<bool>& isVisited, vector<int>& discoveryTime, 
                      vector<int>& lowestTime, vector<vector<Adj>>& adjList, vector<string>& edgeAnswer) {
    isVisited[node] = true;
    discoveryTime[node] = time;
    lowestTime[node] = time;
    for (const auto& adj : adjList[node]) {
        int index = adj.index;
        int adjacentNode = adj.node;
        if (!isVisited[adjacentNode]) {
            depthFirstSearch(adjacentNode, index, time + 1, isVisited, discoveryTime, lowestTime, adjList, edgeAnswer);
            lowestTime[node] = min(lowestTime[node], lowestTime[adjacentNode]);
            if (lowestTime[adjacentNode] > discoveryTime[node]) {
                edgeAnswer[index] = "any";
            }
        } else if (index != callerIndex) {
            lowestTime[node] = min(lowestTime[node], discoveryTime[adjacentNode]);
        }
    }
}

int main() {
    ll numNodes, numEdges;
    cin >> numNodes >> numEdges;

    vector<vector<Edge>> weightEdges(MAX_WEIGHT);
    vector<vector<Adj>> adjList(numNodes + 1);
    vector<int> parent(numNodes + 1);
    vector<bool> isVisited(numNodes + 1);
    vector<int> discoveryTime(numNodes + 1);
    vector<int> lowestTime(numNodes + 1);
    vector<string> edgeAnswer(numEdges + 1, "none");

    int u, v, w, maxWeight = 0;
    for (int i = 1; i <= numEdges; i++) {
        cin >> u >> v >> w;
        Edge e = {u, v, i};
        weightEdges[w].push_back(e);
        maxWeight = max(maxWeight, w);
    }

    for (int i = 1; i <= numNodes; i++) {
        parent[i] = i;
    }

    for (int w = 1; w <= maxWeight; w++) {
        if (weightEdges[w].empty()) {
            continue;
        }

        for (const auto& e : weightEdges[w]) {
            int u = e.node1;
            int v = e.node2;
            int i = e.index;
            if (findParent(u, parent) != findParent(v, parent)) {
                adjList[findParent(u, parent)].push_back({findParent(v, parent), i});
                adjList[findParent(v, parent)].push_back({findParent(u, parent), i});
                isVisited[findParent(u, parent)] = false;
                isVisited[findParent(v, parent)] = false;
                edgeAnswer[i] = "at least one";
            }
        }

        for (const auto& e : weightEdges[w]) {
            int u = e.node1;
            int v = e.node2;
            if (!isVisited[findParent(u, parent)]) {
                depthFirstSearch(findParent(u, parent), -1, 0, isVisited, discoveryTime, lowestTime, adjList, edgeAnswer);
            }
            adjList[findParent(u, parent)].clear();
            adjList[findParent(v, parent)].clear();
            parent[findParent(u, parent)] = findParent(v, parent);
        }
    }

    for (int i = 1; i <= numEdges; i++) {
        cout << edgeAnswer[i] << endl;
    }

    return 0;
}

File: CA4/3.py
#https://chatgpt.com/share/676d9f44-7604-800e-a442-848b2f491c24
#Sayyed Mohammad Hossein Mazhari
#CA4 Q3
from collections import defaultdict

def find_parent(node, parent):
    if node == parent[node]:
        return node
    parent[node] = find_parent(parent[node], parent)
    return parent[node]

def depth_first_search(node, caller_index, time, is_visited, discovery_time, lowest_time, adj_list, edge_answer):
    is_visited[node] = True
    discovery_time[node] = time
    lowest_time[node] = time
    for adj in adj_list[node]:
        index = adj['index']
        adjacent_node = adj['node']
        if not is_visited[adjacent_node]:
            depth_first_search(adjacent_node, index, time + 1, is_visited, discovery_time, lowest_time, adj_list, edge_answer)
            lowest_time[node] = min(lowest_time[node], lowest_time[adjacent_node])
            if lowest_time[adjacent_node] > discovery_time[node]:
                edge_answer[index] = "any"
        elif index != caller_index:
            lowest_time[node] = min(lowest_time[node], discovery_time[adjacent_node])

def main():

    num_nodes, num_edges = map(int, input().split())

    weight_edges = defaultdict(list)
    adj_list = defaultdict(list)
    parent = list(range(num_nodes + 1))
    is_visited = [False] * (num_nodes + 1)
    discovery_time = [0] * (num_nodes + 1)
    lowest_time = [0] * (num_nodes + 1)
    edge_answer = ["none"] * (num_edges + 1)

    max_weight = 0
    for i in range(1, num_edges + 1):
        u, v, w = map(int, input().split())
        weight_edges[w].append({'node1': u, 'node2': v, 'index': i})
        max_weight = max(max_weight, w)

    for w in range(1, max_weight + 1):
        if not weight_edges[w]:
            continue

        for e in weight_edges[w]:
            u, v, i = e['node1'], e['node2'], e['index']
            parent_u = find_parent(u, parent)
            parent_v = find_parent(v, parent)
            if parent_u != parent_v:
                adj_list[parent_u].append({'node': parent_v, 'index': i})
                adj_list[parent_v].append({'node': parent_u, 'index': i})
                is_visited[parent_u] = is_visited[parent_v] = False
                edge_answer[i] = "at least one"

        for e in weight_edges[w]:
            u, v = e['node1'], e['node2']
            parent_u = find_parent(u, parent)
            if not is_visited[parent_u]:
                depth_first_search(parent_u, -1, 0, is_visited, discovery_time, lowest_time, adj_list, edge_answer)

            adj_list[parent_u].clear()
            adj_list[find_parent(v, parent)].clear()
            parent[parent_u] = find_parent(v, parent)

    for i in range(1, num_edges + 1):
        print(edge_answer[i])

if __name__ == "__main__":
    main()
